0006-sync-iteration/README.md

```diff
@@ -0,0 +1,197 @@
+
+A few schema changes are needed to enable validation via inversion:
+
+- the new `prevData` field on `#commit` messages, mentioned above
+- the `repoOp` objects (in `ops` on `#commit` messages) need to include the previous value (CID) for "update" and "delete" actions, so they can be inverted. Note that the record values are not needed, only the CID.
+- in some cases, additional unchanged MST nodes need to be included in `blocks`, so that the tree data structure can be updated
+
+Which additional nodes need to be included, and when? The tautological answer is "whatever blocks are needed for operation inversion". One attempt at a definition is to include blocks which contain the MST keys (paths) directly adjacent to keys added or removed to the tree by any operations. We will provide test vectors covering these situations, and will work towards a more formal definition for the final specifications.
+
+
+## Staying Synchronized: `#sync` event, auto-repair, and account status
+
+The "inductive firehose" process works when each `#commit` is fully verified, in a sequence. What happens if things get off-track? As much as possible, the network should automatically recover from error conditions.
+
+There are two broad categories of errors that can disrupt synchronization:
+
+- **Invalid Commit Message:** the message has invalid CBOR encoding, invalid MST structure, missing blocks, mismatching operation list, etc. Usually caused by *upstream implementation bugs*.
+- **Discontinuous Commits:** re-transmitted or missing commits, `prevData` does not match the actual previous commit, duplicate `rev`, etc. Often caused by *operational issues*, though might also be due to implementation bugs.
+
+A general assumption is that implementation bugs will persist until there is developer intervention, while operational issues are usually transient and can be recovered from automatically. This means that invalid commits are usually dropped by consumers. Some types of discontinuous commits (like retransmission) are also ignored, while others (forks, skipped messages) result in re-synchronization.
+
+A new message type (`#sync`) is added to help with some error conditions. This message is similar to a `#commit` message, in that it contains a single authenticated commit, encoded in a CAR file. It only contains the commit block, not any MST nodes or records (even if the commit did mutate the directory). The semantics of the `#sync` message are: "this is the current state of the repository". When receiving a `#sync`, consumers should check if the `rev` is old (in which case it is ignored) or matches the current state of the repository (a no-op, also ignored). Otherwise, if there is a gap, the consumer needs to resynchronize the repository by fetching a full CAR file using `com.atproto.sync.getRepo`. They should make this request to the same host they are consuming the firehose from (eg, a relay); see the "Sync Boundary" section below for more details.
```

Comment:
I think of this less as "help with error conditions" and more "express operations that do not fit in the semantics of a stream of commits" - which may include error conditions, but also things like account migrations with changes to the MST in between

Reply:
✔️